pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -104,22 +123,50 @@ func adjustZenConfig(k8sClient k8s.Client, es esv1.Elasticsearch, resources node
 // adjustStatefulSetReplicas updates the replicas count in expected according to
 // what is allowed by the upscaleState, that may be mutated as a result.
 func adjustStatefulSetReplicas(
+	k8sClient k8s.Client,
 	upscaleState *upscaleState,
 	actualStatefulSets sset.StatefulSetList,
 	expected appsv1.StatefulSet,
 ) (appsv1.StatefulSet, error) {
-	actual, alreadyExists := actualStatefulSets.GetByName(expected.Name)
 	expectedReplicas := sset.GetReplicas(expected)
-	actualReplicas := sset.GetReplicas(actual)
+
+	actual, alreadyExists := actualStatefulSets.GetByName(expected.Name)
+	actualReplicas := sset.GetReplicas(actual) // 0 if actual is empty
+	if !alreadyExists {
+		// In most cases, if the StatefulSet does not exist we're creating it for the first time (actualReplicas = 0).
+		// However if some Pods matching that missing StatefulSet exist we are likely in a situation where the
+		// StatefulSet is being re-created for volume expansion. In which case we want to take into account
```

Comment:
I'm wondering if it is still true: my understanding is that if a statefulset has been deleted this part of the code is never reached. The reason is that before `HandleUpscaleAndSpecChanges` we are checking if some recreations are in progress:

```golang
	// recreate any StatefulSet that needs to account for PVC expansion
	recreations, err := recreateStatefulSets(d.K8sClient(), d.ES)
	if err != nil {
		return results.WithError(fmt.Errorf("StatefulSet recreation: %w", err))
	}
	if recreations > 0 {
		// Some StatefulSets are in the process of being recreated to handle PVC expansion:
		// it is safer to requeue until the re-creation is done.
		// Otherwise, some operation could be performed with wrong assumptions:
		// the sset doesn't exist (was just deleted), but the Pods do actually exist.
		log.V(1).Info("StatefulSets recreation in progress, re-queueing.",
			"namespace", d.ES.Namespace, "es_name", d.ES.Name, "recreations", recreations)
		return results.WithResult(defaultRequeue)
	}
```

Reply:
Good catch, you're right! We don't need that complicated logic anymore, which is great.